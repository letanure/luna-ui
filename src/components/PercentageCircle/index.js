import React, { useEffect } from 'react'

const radiansFromPercentage = (percentage) => {
  return (percentage * 2) / 100
}

const PercentageCircle = ({ value, lineWidth, size, color, contrastColor, ...restProps }) => {
  let canvas = React.createRef()
  const XPosition = (size+lineWidth)/2
  const YPosition = (size+lineWidth)/2
  const radius = size/2
  const percentageArcEndAngle = radiansFromPercentage(value * 10) * Math.PI
  const canvarSize = size + lineWidth

  useEffect(() => {
    const ctx = canvas.current.getContext("2d")
    if (ctx) {
      // Draws full circle
      ctx.beginPath()
      ctx.strokeStyle = contrastColor
      ctx.lineWidth = lineWidth
      ctx.arc(XPosition, YPosition, radius, 0, 2 * Math.PI)
      ctx.stroke()

      // Draws percentage circle
      ctx.beginPath()
      ctx.lineWidth = lineWidth
      ctx.strokeStyle = color
      ctx.arc(XPosition, YPosition, radius, 0, percentageArcEndAngle)
      ctx.stroke()
    }
  })

  return (
    <canvas {...restProps} width={canvarSize} height={canvarSize} ref={canvas}></canvas>
  )
}

export default PercentageCircle