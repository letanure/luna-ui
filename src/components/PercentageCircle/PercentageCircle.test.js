import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import theme from '../../config/defaultTheme'
import ThemeProvider from '../Theme/index'

import PercentageCircle from './index'

describe('<PercentageCircle />', () => {

  let component

  beforeEach(() => {
    const props = {
      value: 2.5,
      size: 100,
      lineWidth: 8,
      color: '#ffb61f',
      contrastColor: '#f0f0f0'
    }


    component = (
      <PercentageCircle {...props} />
    )
  })

  it('allows us to set value prop', () => {
    const wrapper = mount(component)
    expect(wrapper.props().value).toBe(2.5)
    wrapper.setProps({ value: 5.3 })
    expect(wrapper.props().value).toBe(5.3)
  })

  it('allows us to set value size', () => {
    const wrapper = mount(component)
    expect(wrapper.props().size).toBe(100)
    wrapper.setProps({ size: 200 })
    expect(wrapper.props().size).toBe(200)
  })

  it('allows us to set value lineWidth', () => {
    const wrapper = mount(component)
    expect(wrapper.props().lineWidth).toBe(8)
    wrapper.setProps({ lineWidth: 9 })
    expect(wrapper.props().lineWidth).toBe(9)
  })

  it('allows us to set value color', () => {
    const wrapper = mount(component)
    expect(wrapper.props().color).toBe('#ffb61f')
    wrapper.setProps({ color: '#00ff00' })
    expect(wrapper.props().color).toBe('#00ff00')
  })

  it('allows us to set value lineWidth', () => {
    const wrapper = mount(component)
    expect(wrapper.props().contrastColor).toBe('#f0f0f0')
    wrapper.setProps({ contrastColor: '#00ff00' })
    expect(wrapper.props().contrastColor).toBe('#00ff00')
  })

  test('toMatchSnapshot', () => {
    const tree = renderer
      .create(component)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

})
