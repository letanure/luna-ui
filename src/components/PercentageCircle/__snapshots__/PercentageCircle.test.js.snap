// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PercentageCircle /> toMatchSnapshot 1`] = `
<canvas
  height={108}
  width={108}
/>
`;
